backend/utils/pubsub.js

```diff
@@ -0,0 +1,261 @@
+ *
+ * {number} backlog - The maximum length of the queue of pending connections.
+ * {boolean} clientTracking - Specifies whether or not to track clients.
+ * {Function} handleProtocols - A function which can be used to handle the WebSocket subprotocols.
+ * {number} maxPayload - The maximum allowed message size in bytes.
+ * {string} path - Accept only connections matching this path.
+ * {(boolean|object)} perMessageDeflate - Enables/disables per-message deflate.
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = { ...options.clientHandlers }
+  server.customServerHandlers = { ...options.serverHandlers }
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
+
+  // Create and attach server-side WebSocket event listeners.
+  ;['close', 'connection', 'error', 'headers', 'listening'].forEach((name) => {
+    server.on(name, (...args) => {
+      console.log('[pubsub] Server event:', name)
+      try {
+        const customHandler = server.customServerHandlers['on' + name]
+        const defaultHandler = defaultServerHandlers['on' + name]
+        // Always call the default handler first.
+        if (defaultHandler) {
+          defaultHandler.call(server, ...args)
+        }
+        if (customHandler) {
+          customHandler.call(server, ...args)
+        }
+      } catch (error) {
+        logger(error)
+      }
+    })
+  })
+  // Setup a ping interval if required.
+  if (server.options.pingInterval > 0) {
+    server.pingInterval = setInterval(() => {
+      server.clients.forEach((client) => {
+        if (!client.isAlive) {
+          return client.terminate()
+        }
+        client.isAlive = false
+        client.ping()
+      })
+    }, server.options.pingInterval)
+  }
+  return server
+}
+
+const defaultOptions = {
+  clientTracking: true,
+  maxPayload: 6 * 1024 * 1024,
+  pingInterval: undefined
+}
+
+const generateSocketID = (() => {
+  let counter = 0
+
+  return () => counter++
+})()
+
+// ====== Server socket event handlers ====== //
+
+const defaultServerHandlers = Object.create(null)
+
+/**
+ * Emitted when a connection handshake completes.
+ *
+ * @param socket - The client socket that connected.
+ * @param request - The http GET request sent by the socket.
+ */
+defaultServerHandlers.onconnection = function onconnection (socket, request) {
+  socket.id = generateSocketID()
+  socket.isAlive = true
+  socket.server = this
+  socket.subscriptions = new Set()
+
+  const xForwardedFor = request.headers['x-forwarded-for']
+
+  const remoteAddress = (
+    typeof xForwardedFor === 'string' && xForwardedFor !== ''
+      ? xForwardedFor.split(/\s*,\s*/)[0]
```

Comment:
Thanks for pointing this, we'll definetely have to consider such security issues.

Fow now this remote address value is only used in server console output for debugging during development and testing; it doesn't get stored or exposed.

Reply:
Nevertheless it's currently being presented as *the* client's IP address. If someone were to change the code in the future, using this variable for something that *does* have security impact, it is very unlikely that they will re-evaluate whether the safe-usage assumptions that are made here still hold - and so they will likely introduce a security issue.

If we're going to keep using this 'naive' implementation, then we should at least tag the resulting variable as being unsafe, eg. calling it `unsafeRemoteAddress` instead of just `remoteAddress`. This provides an explicit cue to any future developer working with the code that they need to carefully evaluate the implications of using it.

This also applies more generally to cases of "it's fine that we're doing an unsafe thing because we're only using it safely" - those usage assumptions almost always get invalidated over time, and this is a frequent source of real-world security issues in software. Unsafe things should therefore always be tagged to communicate this.

Reply:
Good points. Therefore, it might be simpler to just stop trying to log the remote address here.

Anyway there is already logging code in `backend/server.js` that does it on every Hapi request. Does the Hapi `request.info.remoteAddress` suffer from the same issues?